### PR TITLE
KAFKA-16026: Send Poll event to the background thread

### PR DESCRIPTION
**Pull request by philipnee**
related to [KAFKA-15818](https://issues.apache.org/jira/browse/KAFKA-15818)

This is a bug in the AsyncKafkaConsumer poll loop that it does not send an event to the network thread to acknowledge user poll. This causes a few issues:
1. Autocommit won't work without user setting the timer
2. the member will just leave the group after rebalance timeout and never able to rejoin.

In this PR, a few subtle changes are made to address this issue
1. Hook up poll event to the AsyncKafkaConsumer#poll. It is only fired once per invocation
2. Upon entering staled state, we need to reset HeartbeatState otherwise we will get an invalid request
3. We will clear and current assignment and remove all assigned partitions once the heartbeat is sent.  See changes in `onHeartbeatRequestSent` 

TODO:
1. We will need to implement the callback invocation logic.  We didn't do it for this PR is because we want to limit to scope of the changes
2. I think FencedException and UnknownMemberIDException: we will also need to reset the HeartbeatState. See [KAFKA-16034](https://issues.apache.org/jira/browse/KAFKA-16034)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java [hunk 1]
```diff
@@ -195,6 +195,7 @@ public NetworkClientDelegate.PollResult poll(long currentTimeMs) {
             NetworkClientDelegate.UnsentRequest request = makeHeartbeatRequest(currentTimeMs, true);
             // We can ignore the leave response because we can join before or after receiving the response.
             heartbeatRequestState.reset();
+            heartbeatState.reset();
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:198]: We might need to reset this at other places where it requires to rejoin.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java [hunk 2]
```diff
@@ -184,7 +184,9 @@ public NetworkClientDelegate.PollResult poll(long currentTimeMs) {
             return NetworkClientDelegate.PollResult.EMPTY;
         }
         pollTimer.update(currentTimeMs);
-        if (pollTimer.isExpired()) {
+        // If the poll timer expires during reconciliation, we need to wait till the reconciliation completes before
+        // sending another leave group.
```
Comment (cadonna) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: Could you please elaborate why we need to wait? It is not clear to me, because I thought that we can leave the group whenever we want.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java [hunk 3]
```diff
@@ -672,7 +672,11 @@ public void onHeartbeatRequestSent() {
         MemberState state = state();
         if (isStaled()) {
             log.debug("Member {} is staled and is therefore leaving the group.  It will rejoin upon the next poll.", memberEpoch);
-            transitionToJoining();
+            // clear the current assignment and subscription, and trigger rebalance listener on the next poll
+            invokeOnPartitionsRevokedCallback(subscriptions.assignedPartitions()).whenComplete((r, e) -> {
```
Comment (cadonna) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java]: Is it correct to call `onPartitionRevoked()` here? Technically, the member is not a member of the group anymore. I am not sure it should be allowed to do anything group-related like committing offsets which is one of the main purposes of `onPartitionRevoked()`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java [hunk 4]
```diff
@@ -728,7 +732,6 @@ public boolean shouldSkipHeartbeat() {
     @Override
     public void transitionToStaled() {
         memberEpoch = ConsumerGroupHeartbeatRequest.LEAVE_GROUP_MEMBER_EPOCH;
-        currentAssignment.clear();
```
Comment (cadonna) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:757]: Why do we not need to remove the current assignment if we leave the group? Is this because of `updateSubscription()` on line 677?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java [hunk 5]
```diff
@@ -184,7 +184,9 @@ public NetworkClientDelegate.PollResult poll(long currentTimeMs) {
...
+        if (pollTimer.isExpired() && membershipManager.state() == MemberState.STABLE) {
```
Comment (cadonna) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: Is there a unit test that tests the behavior when this condition is not satisfied?

**Review by cadonna — COMMENTED** [review]
Thanks for the PR, @philipnee !

Here my feedback.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java (hunk 2 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: if the member is reconciling its assignment, we need to wait for the process to finish before revoking all assignments, then leave the group.

**Comment by cadonna** [comment]
Test `testEnsurePollEventSentOnConsumerPoll()` fails deterministically in all CI builds. That means that something is wrong either with the test you added or with the production code that is tested. Please ensure that at least the tests you added or modified or are affected by your change are successful by running them locally. Otherwise, we review code that obviously needs changes and we end up cycling over the PR more than needed.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java (hunk 3 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java]: when the member becomes staled, it is still part of the group because it hasn't left yet.  however, the background thread will need to leave the group and revoke all of its assignment because we assume the consumer is not being used - this is the logic for max.poll.interval.ms.  since we are leaving the group, we will need to revoke all assignments.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java (hunk 4 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:757]: correct

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java (hunk 5 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: good call - will add one.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java (hunk 3 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java]: Isn't revocation something that the broker tells the member to because of rebalancing? In this case it is the member that decides to leave the group because it exceeded the poll interval. It is similar to close. I do not think we revoke partitions in close, do we?

Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java]: OK, I just saw https://issues.apache.org/jira/browse/KAFKA-15696

Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java]: Revocation at `close()`-time seems legit.

Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java]: I discussed this offline with @dajac and we think that since exceeding the poll interval is something unexpected we should rather call `onPartitionLost()` rather than `onPartitionRevoked()`.

Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java]: I agree.  We had a bit of discussion around this.  I think whether calling onPartitionLost/Revoke is up for discussion. Perhaps we limit the scope of the change to just clearing the assignment subscription w/o invoking the callbacks.

Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java]: Yes, good idea! I agree!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java (hunk 2 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: the change was removed - i think we don't need to make a special case for reconciliation.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java (hunk 4 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:757]: See the TODO on line 676.  For now, I think we can directly clear the currentAssignment and remove owned partitions.  In the future, we will need to wait for the callback to complete before proceed with joining.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java [hunk 6]
```diff
@@ -136,7 +135,7 @@ private void process(final PollApplicationEvent event) {
         }
 
         requestManagers.commitRequestManager.ifPresent(m -> m.updateAutoCommitTimer(event.pollTimeMs()));
-        requestManagers.heartbeatRequestManager.ifPresent(HeartbeatRequestManager::resetPollTimer);
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java:140]: turned out this is a bug - the timer's current time was not updated.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java [hunk 7]
```diff
@@ -385,6 +386,7 @@ public void testWakeupAfterEmptyFetch() {
         assertDoesNotThrow(() -> consumer.poll(Duration.ZERO));
     }
 
+    @Test
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java:388]: I have no idea why this was removed.  It could be me.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java [hunk 8]
```diff
@@ -660,14 +672,14 @@ private HeartbeatRequestManager createHeartbeatRequestManager() {
         subscriptions = mock(SubscriptionState.class);
         membershipManager = mock(MembershipManager.class);
         backgroundEventHandler = mock(BackgroundEventHandler.class);
-        heartbeatState = new HeartbeatRequestManager.HeartbeatState(subscriptions, membershipManager, maxPollIntervalMs);
-        heartbeatRequestState = new HeartbeatRequestManager.HeartbeatRequestState(
+        heartbeatState = spy(new HeartbeatRequestManager.HeartbeatState(subscriptions, membershipManager, maxPollIntervalMs));
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java]: the purpose of spying was to make sure a certain method was invoked. I think this is an acceptable use case, do you agree?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java [hunk 9]
```diff
@@ -288,7 +288,7 @@ public CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> addOffsetFetchR
     }
 
     public void updateAutoCommitTimer(final long currentTimeMs) {
-        this.autoCommitState.ifPresent(t -> t.ack(currentTimeMs));
+        this.autoCommitState.ifPresent(t -> t.updateTimer(currentTimeMs));
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:309]: switching to updateTimer for a more descriptive name.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java (hunk 1 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:198]: One example would be fenced exception - we need to reset before sending a joining heartbeat.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java [hunk 10]
```diff
@@ -234,7 +237,8 @@ public long maximumTimeToWait(long currentTimeMs) {
      * When consumer polls, we need to reset the pollTimer.  If the poll timer has expired, we rejoin only when the
      * member is in the {@link MemberState#UNSUBSCRIBED} state.
      */
-    public void resetPollTimer() {
+    public void resetPollTimer(final long pollMs) {
+        pollTimer.update(pollMs);
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:239]: note: pollMs is the time user invoke consumer#poll.  This is better than taking time.millisecond() because there could be some discrepancy.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 11]
```diff
@@ -169,7 +169,6 @@ class PlaintextConsumerTest extends BaseConsumerTest {
       startingTimestamp = startingTimestamp)
   }
 
-  // TODO: enable this test for the consumer group protocol when KAFKA-16008 has been fixed.
```
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:172]: Why did you remove this comment but not enable the test?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 12]
```diff
@@ -196,7 +195,6 @@ class PlaintextConsumerTest extends BaseConsumerTest {
     assertEquals(1, listener.callsToRevoked)
   }
 
-  // TODO: enable this test for the consumer group protocol when KAFKA-16009 has been fixed.
```
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:199]: Why did you remove this comment but not enable the test?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java [hunk 13]
```diff
@@ -234,7 +235,8 @@ public long maximumTimeToWait(long currentTimeMs) {
      * When consumer polls, we need to reset the pollTimer.  If the poll timer has expired, we rejoin only when the
      * member is in the {@link MemberState#UNSUBSCRIBED} state.
```
Comment (cadonna) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: Is the state here correct?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java (hunk 8 above)
Comment (cadonna) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java]: No, I don't.  Why not passing mocks to this method.

**Comment by philipnee** [comment]
hi @cadonna - Thanks for putting time into the review.  I've addressed your comments.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java [hunk 14]
```diff
@@ -673,6 +673,9 @@ public void onHeartbeatRequestSent() {
         MemberState state = state();
         if (isStaled()) {
             log.debug("Member {} is staled and is therefore leaving the group.  It will rejoin upon the next poll.", memberEpoch);
+            // TODO: Integrate partition revocation/loss callback
+            // Clear the current assignment and subscribed partitions because the member has left the group
+            updateSubscription(new TreeSet<>(TOPIC_ID_PARTITION_COMPARATOR), true);
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java]: Should we rather call this before `transitionTo(MemberState.STALED);`?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java [hunk 15]
```diff
@@ -528,13 +529,18 @@ public void testPollTimerExpiration() {
         heartbeatRequestManager = createHeartbeatRequestManager();
         when(coordinatorRequestManager.coordinator()).thenReturn(Optional.of(new Node(1, "localhost", 9999)));
         when(membershipManager.shouldSkipHeartbeat()).thenReturn(false);
+        when(membershipManager.state()).thenReturn(MemberState.STABLE);
+        doNothing().when(membershipManager).transitionToStaled();
```
Comment (cadonna) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java]: I do not think, you need this. Should also work without.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java (hunk 14 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java]: Do you mean in `transitionToStaled()`? That makes sense to me.

**Review by cadonna — COMMENTED** [review]
Thanks for the updates, @philipnee !

I have just a minor comment.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java (hunk 14 above)
Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java]: Indeed. I find this very confusing. `STALED` is not a word. I'm not sure whether it means `STALE` or `STALLED`, although I think it's the former. If we want to make a verb for this, I suggest `GONE_STALE`. But really, I think that `transitionToStale` is probably best, along with changing to `MemberState.STALE`.

Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java]: > Do you mean in `transitionToStaled()`? That makes sense to me.

Yes.

Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java]: Thanks, I originally thought we wanted to abandon the partitions after sending the leave group.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java [hunk 16]
```diff
@@ -191,10 +191,11 @@ public NetworkClientDelegate.PollResult poll(long currentTimeMs) {
                 "either by increasing max.poll.interval.ms or by reducing the maximum size of batches " +
                 "returned in poll() with max.poll.records.");
             // This should trigger a heartbeat with leave group epoch
-            membershipManager.transitionToStaled();
+            membershipManager.transitionToStale();
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:194]: transitionTo**Stale**

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java [hunk 17]
```diff
@@ -679,4 +683,23 @@ private HeartbeatRequestManager createHeartbeatRequestManager() {
             heartbeatRequestState,
             backgroundEventHandler);
     }
+
+    private HeartbeatRequestManager createHeartbeatRequestManager(
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:688]: turned out we are using testBuilder in this test again.  So we might want to refactor the test and turn everything to mock.

**Comment by philipnee** [comment]
@cadonna @dajac @AndrewJSchofield  - Thanks a lot for the reviews.  I've made changes accordingly.

**Review by AndrewJSchofield — APPROVED** [review]
The addition of the poll event is a good improvement. lgtm.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java [hunk 18]
```diff
@@ -525,16 +526,37 @@ public void testHeartbeatState() {
 
     @Test
     public void testPollTimerExpiration() {
-        heartbeatRequestManager = createHeartbeatRequestManager();
+        coordinatorRequestManager = mock(CoordinatorRequestManager.class);
+        membershipManager = mock(MembershipManager.class);
+        heartbeatState = mock(HeartbeatRequestManager.HeartbeatState.class);
+        heartbeatRequestState = spy(new HeartbeatRequestManager.HeartbeatRequestState(
```
Comment (cadonna) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:533]: Why is there still this spy?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java (hunk 14 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java]: Is it correct that we have this call now in two places?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java (hunk 18 above)
Comment (cadonna) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:533]: If it is too much work to change, let's change it later and merge it as it is.

**Review by cadonna — APPROVED** [review]
LGTM!

Thanks, @philipnee !

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java (hunk 18 above)
Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:533]: it is there to verify the reset call and we also need a real object to ensure the manager can send a request.  If we mock it, then we have to mock the canSendRequest call, then it is pointless to verify the request returned from poll because it will always return one.

**Comment by philipnee** [comment]
hmm @cadonna - seems like there are some build stability issue.  my other PR also failed after 8hr (it is completely unrelated to the consumer refactor) `PR-15023`

**Comment by cadonna** [comment]
Don't worry! We will take care of restarting the builds.

**Comment by philipnee** [comment]
@dajac - thanks for merging the changes.
